akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+      def offer(id: Long, elem: Any): Unit
+      def poll(id: Long): AnyRef
+      def remove(id: Long): Unit
+    }
+
+    class PartitionQueueImpl extends PartitionQueue {
+      private val queues1 = new AtomicReferenceArray[Queue[Any]](FixedQueues)
+      private val queues2 = new ConcurrentHashMap[Long, Queue[Any]]
+      private val _size = new AtomicInteger
+
+      def init(id: Long): Unit = {
+        if (id < FixedQueues)
+          queues1.set(id.toInt, Queue.empty)
+        else
+          queues2.put(id, Queue.empty)
+      }
+
+      def size: Int = _size.get
+
+      def isEmpty(id: Long): Boolean = {
+        val queue =
+          if (id < FixedQueues) queues1.get(id.toInt)
+          else queues2.get(id)
+        queue.isEmpty
+      }
+
+      def nonEmpty(id: Long): Boolean = !isEmpty(id)
+
+      def offer(id: Long, elem: Any): Unit = {
+        @tailrec def offer1(): Unit = {
+          val i = id.toInt
+          val queue = queues1.get(i)
+          if (queue eq null)
+            throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+          if (queues1.compareAndSet(i, queue, queue.enqueue(elem)))
+            _size.incrementAndGet()
+          else
+            offer1() // CAS failed, retry
+        }
+
+        @tailrec def offer2(): Unit = {
+          val queue = queues2.get(id)
+          if (queue eq null)
+            throw new IllegalArgumentException(s"Invalid stream identifier: $id")
+          if (queues2.replace(id, queue, queue.enqueue(elem))) {
+            _size.incrementAndGet()
+          } else
+            offer2() // CAS failed, retry
+        }
+
+        if (id < FixedQueues) offer1() else offer2()
+      }
+
+      def poll(id: Long): AnyRef = {
+        @tailrec def poll1(): AnyRef = {
+          val i = id.toInt
+          val queue = queues1.get(i)
+          if ((queue eq null) || queue.isEmpty) null
+          else if (queues1.compareAndSet(i, queue, queue.tail)) {
+            _size.decrementAndGet()
+            queue.head.asInstanceOf[AnyRef]
+          } else
+            poll1() // CAS failed, try again
+        }
+
+        @tailrec def poll2(): AnyRef = {
+          val queue = queues2.get(id)
+          if ((queue eq null) || queue.isEmpty) null
+          else if (queues2.replace(id, queue, queue.tail)) {
+            _size.decrementAndGet()
+            queue.head.asInstanceOf[AnyRef]
+          } else
+            poll2() // CAS failed, try again
+        }
+
+        if (id < FixedQueues) poll1() else poll2()
+      }
+
+      def remove(id: Long): Unit = {
+        (if (id < FixedQueues) queues1.getAndSet(id.toInt, null)
+        else queues2.remove(id)) match {
+          case null  ⇒
+          case queue ⇒ _size.addAndGet(-queue.size)
+        }
+      }
+
+    }
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class PartitionHub[T](partitioner: () ⇒ (Array[Long], T) ⇒ Long, startAfterNbrOfConsumers: Int, bufferSize: Int)
```

Comment:
Good point, that would be useful. I think we can expose the queue sizes for each consumer, which should be a good indicator. I'll look into that tomorrow.